File: nav/js/frames.js
// Coordinate conventions - JS port of src/navigation/frames.py (keep in sync).
// ENU local tangent plane in metres; psi = heading CCW from East (rad); compass = deg clockwise from North.

const WGS84_A = 6378137.0;
const WGS84_E2 = 6.69437999014e-3;
const DEG = Math.PI / 180;

export function wrapPi(a) {
  const r = (a + Math.PI) % (2 * Math.PI);
  return (r < 0 ? r + 2 * Math.PI : r) - Math.PI;
}

export const compassToPsi = (courseDeg) => wrapPi(Math.PI / 2 - courseDeg * DEG);

export function psiToCompass(psi) {
  const d = (Math.PI / 2 - psi) / DEG;
  return ((d % 360) + 360) % 360;
}

function ecef(latDeg, lonDeg, h) {
  const lat = latDeg * DEG, lon = lonDeg * DEG;
  const sl = Math.sin(lat), cl = Math.cos(lat);
  const n = WGS84_A / Math.sqrt(1 - WGS84_E2 * sl * sl);
  return [(n + h) * cl * Math.cos(lon), (n + h) * cl * Math.sin(lon), (n * (1 - WGS84_E2) + h) * sl];
}

export class LocalTangentPlane {
  constructor(lat0, lon0, h0 = 0) {
    this.lat0 = lat0; this.lon0 = lon0; this.h0 = h0;
    const lat = lat0 * DEG, lon = lon0 * DEG;
    this.sl = Math.sin(lat); this.cl = Math.cos(lat);
    this.so = Math.sin(lon); this.co = Math.cos(lon);
    [this.x0, this.y0, this.z0] = ecef(lat0, lon0, h0);
  }

  toEnu(lat, lon) {
    const [x, y, z] = ecef(lat, lon, this.h0);
    const dx = x - this.x0, dy = y - this.y0, dz = z - this.z0;
    const e = -this.so * dx + this.co * dy;
    const n = -this.sl * this.co * dx - this.sl * this.so * dy + this.cl * dz;
    return [e, n];
  }

  toGeodetic(e, n) {
    const x = this.x0 - this.so * e - this.sl * this.co * n;
    const y = this.y0 + this.co * e - this.sl * this.so * n;
    const z = this.z0 + this.cl * n;
    const lon = Math.atan2(y, x);
    const p = Math.hypot(x, y);
    let lat = Math.atan2(z, p * (1 - WGS84_E2));
    for (let i = 0; i < 3; i++) {
      const nn = WGS84_A / Math.sqrt(1 - WGS84_E2 * Math.sin(lat) ** 2);
      const h = p / Math.cos(lat) - nn;
      lat = Math.atan2(z, p * (1 - WGS84_E2 * nn / (nn + h)));
    }
    return [lat / DEG, lon / DEG];
  }
}

File: src/navigation/frames.py
"""
Coordinate-frame conventions shared by every IDR module.

Navigation frame : local East-North-Up (ENU) tangent plane, metres.
Heading psi      : radians, counter-clockwise from East (the ENU yaw angle).
Compass course   : degrees, clockwise from North, in [0, 360) - what GNSS receivers,
                   OSM bearings and the NavigationState output use.

    psi = pi/2 - radians(course)        course = degrees(pi/2 - psi) mod 360
"""
import math

import numpy as np

WGS84_A = 6378137.0
WGS84_E2 = 6.69437999014e-3


def wrap_pi(angle):
    """Wraps an angle (scalar or array) to [-pi, pi)."""
    return (np.asarray(angle) + math.pi) % (2.0 * math.pi) - math.pi if isinstance(angle, np.ndarray) \
        else (angle + math.pi) % (2.0 * math.pi) - math.pi


def compass_to_psi(course_deg):
    return wrap_pi(math.pi / 2.0 - np.radians(course_deg)) if isinstance(course_deg, np.ndarray) \
        else wrap_pi(math.pi / 2.0 - math.radians(course_deg))


def psi_to_compass(psi_rad):
    if isinstance(psi_rad, np.ndarray):
        return np.degrees(math.pi / 2.0 - psi_rad) % 360.0
    return math.degrees(math.pi / 2.0 - psi_rad) % 360.0


class LocalTangentPlane:
    """
    WGS84 geodetic <-> local ENU converter anchored at (lat0, lon0).
    Uses the exact ECEF rotation, so it stays accurate over the 50-100 km extents
    of the IO-VNBD trips (the flat 111320 m/deg approximation does not).
    """

    def __init__(self, lat0_deg, lon0_deg, h0=0.0):
        self.lat0 = float(lat0_deg)
        self.lon0 = float(lon0_deg)
        self.h0 = float(h0)
        lat, lon = math.radians(self.lat0), math.radians(self.lon0)
        self._sl, self._cl = math.sin(lat), math.cos(lat)
        self._so, self._co = math.sin(lon), math.cos(lon)
        self._x0, self._y0, self._z0 = self._ecef(self.lat0, self.lon0, self.h0)

    @staticmethod
    def _ecef(lat_deg, lon_deg, h):
        lat, lon = np.radians(lat_deg), np.radians(lon_deg)
        sl, cl = np.sin(lat), np.cos(lat)
        n = WGS84_A / np.sqrt(1.0 - WGS84_E2 * sl * sl)
        return (n + h) * cl * np.cos(lon), (n + h) * cl * np.sin(lon), (n * (1.0 - WGS84_E2) + h) * sl

    def to_enu(self, lat_deg, lon_deg, h=None):
        """Returns (east, north) in metres. Accepts scalars or arrays."""
        h = self.h0 if h is None else h
        x, y, z = self._ecef(lat_deg, lon_deg, h)
        dx, dy, dz = x - self._x0, y - self._y0, z - self._z0
        e = -self._so * dx + self._co * dy
        n = -self._sl * self._co * dx - self._sl * self._so * dy + self._cl * dz
        if np.ndim(e) == 0:
            return float(e), float(n)
        return e, n

    def to_geodetic(self, east, north):
        """Inverse of to_enu for the horizontal plane (up = 0). Accurate to mm over 100 km."""
        dx = -self._so * east - self._sl * self._co * north
        dy = self._co * east - self._sl * self._so * north
        dz = self._cl * north
        x, y, z = self._x0 + dx, self._y0 + dy, self._z0 + dz
        lon = np.arctan2(y, x)
        p = np.hypot(x, y)
        lat = np.arctan2(z, p * (1.0 - WGS84_E2))
        for _ in range(3):
            n = WGS84_A / np.sqrt(1.0 - WGS84_E2 * np.sin(lat) ** 2)
            h = p / np.cos(lat) - n
            lat = np.arctan2(z, p * (1.0 - WGS84_E2 * n / (n + h)))
        lat, lon = np.degrees(lat), np.degrees(lon)
        if np.ndim(lat) == 0:
            return float(lat), float(lon)
        return lat, lon
